File: hm3/problem2.py
import time
from multiprocessing import Pool
import os
import random

def randomString():
    letters = [random.choice("ABCDEFGHIJKLMNOPQRSTUVWXYZ") for i in range(995)]
    pos = random.randint(0, 994)
    letters = letters[:pos] + ["A", "B", "C", "B", "A"] + letters[pos:]
    return "".join(letters)

def file_generator():
    for i in range(1, 101):
        file = open(str(i) + ".txt", "w")
        randomTxt = randomString()
        a = file.write(randomTxt)
        file.close()

def has_five_letter_palindrome(string):
    for i in range(len(string)-4):
        substring = string[i:i+5]
        if substring == substring[::-1]:
            return True
    return False

def examine_strings(string):
    for i in range(len(string)):
        for j in range(i+1, len(string)+1):
            if has_five_letter_palindrome(string[i:j]):
                return True
    return False

def process_file(filename):
    with open(filename, 'r') as file:
        data = file.read()
    found = examine_strings(data)
    if found:
        return 1
    else:
        return 0

if __name__ == '__main__':

    
    count = 0
    start = time.time()
    for i in range(1, 101):
       count += process_file('palindrome_txts/{}.txt'.format(i))
    end = time.time()
    sequential_time = end - start

    print("Palindromes Found:" , count)
    print("Sequential Elapsed Time:" , sequential_time)

    start = time.time()
    pool = Pool(os.cpu_count())
    count = sum(pool.map(process_file, ['palindrome_txts/{}.txt'.format(i) for i in range(1, 101)]))
    end = time.time()
    parallel_time = end - start
    speedup = sequential_time/parallel_time

    print("Palindromes Found:", count)
    print("ParallelElapsed Time:", parallel_time)

    print("Speed Up: ", speedup)
    print("Efficiency : ", speedup/os.cpu_count())

File: hm3/problem2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <chrono>

using namespace std;

bool has_five_letter_palindrome(string str)
{
    for (int i = 0; i < str.length() - 4; i++)
    {
        string substring = str.substr(i, 5);
        if (substring == string(substring.rbegin(), substring.rend()))
        {
            return true;
        }
    }
    return false;
}

bool examine_strings(string str)
{
    for (int i = 0; i < str.length(); i++)
    {
        for (int j = i+1; j <= str.length(); j++)
        {
            if (has_five_letter_palindrome(str.substr(i, j-i)))
            {
                return true;
            }
        }
    }
    return false;
}

int main()
{
    int count = 0;
    auto start = chrono::high_resolution_clock::now();
    for (int i = 1; i <= 100; i++)
    {
        string filename = "palindrome_txts/" + to_string(i) + ".txt";
        ifstream file(filename);
        string data((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
        bool found = examine_strings(data);
        if (found)
        {
            count++;
        }
    }
    auto end = chrono::high_resolution_clock::now();
    double sequential_time = chrono::duration_cast<chrono::milliseconds>(end - start).count() / 1000.0;

    cout << "Palindromes Detected: " << count << endl;
    cout << "Elapsed Time: " << sequential_time << "s" << endl;

    return 0;
}
